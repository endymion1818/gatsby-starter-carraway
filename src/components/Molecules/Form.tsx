import React, { useEffect, useRef, useState } from 'react'
import styled from 'styled-components'
import ButtonStyles from '../Atoms/ButtonStyles'
import { borderradius, size } from '../tokens'

interface IErrors {
  email: string[]
  yourname: string[]
  phone: string[]
}

const Label = styled.label`
  display: block;
  max-width: 100%;
  margin-bottom: 5px;
  font-size: 120%;
  font-weight: bold;
`

const Input = styled.input`
  display: block;
  width: 80%;
  height: 34px;
  padding: 6px 12px;
  font-size: 14px;
  line-height: 1.42857143;
  color: #555;
  background-color: #fff;
  background-image: none;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075);
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;

  &:focus {
    border-color: #66afe9;
    outline: 0;
    box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075), 0 0 8px rgba(102, 175, 233, 0.6);
  }
  &::-moz-placeholder {
    color: #999;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    color: #999;
  }
  &::-webkit-input-placeholder {
    color: #999;
  }
  &::-ms-expand {
    background-color: transparent;
    border: 0;
  }
  &[disabled],
  &[readonly],
  fieldset[disabled] & {
    background-color: #eee;
    opacity: 1;
  }
  &[disabled],
  fieldset[disabled] .form-control {
    cursor: not-allowed;
  }
`

const Button = styled.button`
  ${ButtonStyles}
`

const FormGroup = styled.div`
  margin-bottom: 15px;
`

const Error = styled.div`
  & > * {
    margin-top: ${size.single};
    border: 1px solid red;
    background-color: pink;
    padding: ${size.single};
    border-radius: ${borderradius.medium};
  }
`

function validateEmail(email: string) {
  const errors = [] as string[]
  if (email.length === 0) {
    errors.push("email can't be empty")
  }
  if (!email.includes('@')) {
    errors.push('Email should contain @')
  }
  return errors
}
function validatePhone(phone: string) {
  const errors = [] as string[]
  if (phone.length === 0) {
    errors.push("Can't be empty")
  }
  return errors
}
function validateName(yourname: string) {
  const errors = [] as string[]
  if (yourname.length === 0) {
    errors.push("Can't be empty")
  }
  return errors
}

function hasErrors(errors: IErrors) {
  return errors.email.length > 0 || errors.yourname.length > 0 || errors.phone.length > 0
}

function Form() {
  const email = useRef<HTMLInputElement | null>(null)
  const yourname = useRef<HTMLInputElement | null>(null)
  const phone = useRef<HTMLInputElement | null>(null)

  const [errors, setErrors] = useState<IErrors>({
    email: [],
    yourname: [],
    phone: [],
  })

  const [isSubmitted, setIsSubmitted] = useState(false)

  useEffect(() => {
    // if JS is disabled, the user can still use the form
    // otherwise, set initial state to false
    setIsSubmitted(false)
  }),
    // tslint:disable-next-line
    []

  useEffect(() => {
    if (hasErrors(errors)) {
      setIsSubmitted(false)
    }
    setIsSubmitted(true)
  }, [isSubmitted, errors])

  function handleEmailChange(e: React.ChangeEvent<HTMLInputElement>) {
    const value = e.target.value
    setErrors((prevState) => {
      return {
        ...prevState,
        email: validateEmail(value),
      }
    })
  }
  function handleNameChange(e: React.ChangeEvent<HTMLInputElement>) {
    const value = e.target.value
    setErrors((prevState) => {
      return {
        ...prevState,
        yourname: validateName(value),
      }
    })
  }
  function handlePhoneChange(e: React.ChangeEvent<HTMLInputElement>) {
    const value = e.target.value
    setErrors((prevState) => {
      return {
        ...prevState,
        phone: validatePhone(value),
      }
    })
  }

  function handleSubmit() {
    setErrors((prevState) => {
      return {
        ...prevState,
        email: validateEmail(email.current!.value),
        phone: validatePhone(phone.current!.value),
        yourname: validateName(yourname.current!.value),
      }
    })
  }

  return (
    <form onSubmit={handleSubmit} method="POST" action="https://formspree.io/rxd_david@gmail.com">
      <FormGroup>
        <Label htmlFor="yourname">Your Name</Label>
        <Input id="yourname" required ref={yourname} onChange={handleNameChange} />
        <Error>
          {errors.yourname.map((error) => (
            <div key={error}>{error}</div>
          ))}
        </Error>
      </FormGroup>

      <FormGroup>
        <Label htmlFor="email">Email</Label>
        <Input type="email" id="email" ref={email} onChange={handleEmailChange} required />
        <Error>
          {errors.email.map((error) => (
            <div key={error}>{error}</div>
          ))}
        </Error>
      </FormGroup>

      <FormGroup>
        <Label htmlFor="phone">
          Phone <small>(optional)</small>
        </Label>
        <Input id="phone" type="tel" ref={phone} onChange={handlePhoneChange} />
        <Error>
          {errors.phone.map((error) => (
            <div key={error}>{error}</div>
          ))}
        </Error>
      </FormGroup>
      <Button disabled={hasErrors(errors)}>Sign Up</Button>
    </form>
  )
}

export default Form
