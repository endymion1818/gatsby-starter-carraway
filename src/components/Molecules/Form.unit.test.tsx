import { cleanup, render } from '@testing-library/react'
import React from 'react'
import Form from './Form'

beforeEach(cleanup)

it('Should render', () => {
  const { container } = render(<Form />)
  expect(container).toMatchSnapshot()
})
