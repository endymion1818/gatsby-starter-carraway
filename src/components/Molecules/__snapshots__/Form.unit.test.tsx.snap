// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render 1`] = `
.c1 {
  display: block;
  max-width: 100%;
  margin-bottom: 5px;
  font-size: 120%;
  font-weight: bold;
}

.c2 {
  display: block;
  width: 80%;
  height: 34px;
  padding: 6px 12px;
  font-size: 14px;
  line-height: 1.42857143;
  color: #555;
  background-color: #fff;
  background-image: none;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075);
  -webkit-transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
}

.c2:focus {
  border-color: #66afe9;
  outline: 0;
  box-shadow: inset 0 1px 1px rgba(0,0,0,0.075),0 0 8px rgba(102,175,233,0.6);
}

.c2::-moz-placeholder {
  color: #999;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #999;
}

.c2::-webkit-input-placeholder {
  color: #999;
}

.c2::-ms-expand {
  background-color: transparent;
  border: 0;
}

.c2[disabled],
.c2[readonly],
fieldset[disabled] .c2 {
  background-color: #eee;
  opacity: 1;
}

.c2[disabled],
.c2 fieldset[disabled] .form-control {
  cursor: not-allowed;
}

.c4 {
  display: inline-block;
  border: none;
  padding: 1rem 2rem;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #DCDBE2;
  color: #000080;
  font-size: 1rem;
  cursor: pointer;
  text-align: center;
  -webkit-transition: background 250ms ease-in-out,-webkit-transform 150ms ease;
  -webkit-transition: background 250ms ease-in-out,transform 150ms ease;
  transition: background 250ms ease-in-out,transform 150ms ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  border-radius: 0.5rem;
}

.c4:hover,
.c4:focus {
  background: #DCDBE2;
}

.c4:focus {
  outline: 1px solid #fff;
  outline-offset: -4px;
}

.c4:active {
  -webkit-transform: scale(0.99);
  -ms-transform: scale(0.99);
  transform: scale(0.99);
}

.c4:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c0 {
  margin-bottom: 15px;
}

.c3 > * {
  margin-top: 1rem;
  border: 1px solid red;
  background-color: pink;
  padding: 1rem;
  border-radius: 0.5rem;
}

<div>
  <form
    action="https://formspree.io/rxd_david@gmail.com"
    method="POST"
  >
    <div
      class="c0"
    >
      <label
        class="c1"
        for="yourname"
      >
        Your Name
      </label>
      <input
        class="c2"
        id="yourname"
        required=""
      />
      <div
        class="c3"
      />
    </div>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="email"
      >
        Email
      </label>
      <input
        class="c2"
        id="email"
        required=""
        type="email"
      />
      <div
        class="c3"
      />
    </div>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="phone"
      >
        Phone 
        <small>
          (optional)
        </small>
      </label>
      <input
        class="c2"
        id="phone"
        type="tel"
      />
      <div
        class="c3"
      />
    </div>
    <button
      class="c4"
    >
      Sign Up
    </button>
  </form>
</div>
`;
